### PR TITLE
Cleanup tracks based on polygon

### PR DESCRIPTION
**Pull request by niravi**
I have only implemented the portion which deletes tracks inside a polygon so far - since it was much simpler

I'm thinking the easiest solution (though not necessarily most performant) for deleting features outside a polygon would be to perform a spatial query and then use the GUIDs in the results of the spatial query to build a WHERE clause and then delete using the same technique:

i.e. tracks_layer.delete_features('global_id NOT in `guid1` OR `guid2`.......`guid1000`') but not sure if this will run into query string limit

### CODE REVIEW

File: scripts/polygon_cleanup_tracks.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    logger.info("Getting polygon layer")
+    try:
+        layer = FeatureLayer(url=args.layer_url, gis=gis)
+        json = layer._lyr_json
+    except Exception as e:
+        logger.info(e)
+        logger.info("Layer could not be found based on given input. Please check your parameters again. Exiting the script")
+        sys.exit(0)
+    
+    features = layer.query(where=args.where, out_sr=4326).features
```
Comment (apulverizer) [inline scripts/polygon_cleanup_tracks.py]: you should use a projected coordinate system like 3857 web mercator

File: scripts/polygon_cleanup_tracks.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    if len(features) > 0:
+        geometries = [feature.geometry for feature in features]
+        union_geometry = geometry.union(spatial_ref=4326,geometries=geometries,gis=gis)
+        intersect_filter = geometry.filters.intersects(union_geometry, sr=4326)
+        if args.inside:
+            tracks_layer.delete_features(geometry_filter=intersect_filter)
```
Comment (apulverizer) [inline scripts/polygon_cleanup_tracks.py]: Probably should check the result here in case it fails

**Review by apulverizer — CHANGES_REQUESTED** [review]
I wonder if we can reverse the order of vertices to do the inverse... like that Cosmos issue we had.

I don't think we should do the long query, that's definitely going to cause problems.

File: scripts/polygon_cleanup_tracks.py (hunk 1 above)
Comment (niravi) [reply scripts/polygon_cleanup_tracks.py]: I thought the LTS was 4326? It appears to be working when I use this, but I get a weird JSON error when I use 3857 (haven't dug too far)

**Comment by niravi** [comment]
@apulverizer used the inverse (counterclockwise geometry) - seems to be working pretty well. I used the "Parks - Areas" layer in the "Parks" flc for testing

File: scripts/polygon_cleanup_tracks.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+        "This sample allows cleanup of track points from a tracks layer based on a spatial relationship to polygon geometry")
+    parser.add_argument('-u', dest='username', help="The username to authenticate with", required=True)
+    parser.add_argument('-p', dest='password', help="The password to authenticate with", required=True)
+    parser.add_argument('-org', dest='org_url', help="The url of the org/portal to use", required=True)
+    # Parameters for tracker
+    parser.add_argument('-layer-url', dest='layer_url',
+                        help="The feature service URL for your layer with the geometry you want to use to delete track points",
+                        required=True)
+    parser.add_argument('-where', dest='where', help="Query conditions for polygons you want to use in your cleanup. Defaults to all (1=1)", default="1=1")
+    parser.add_argument('-delete-inside', dest='inside', help="If true, delete the tracks inside the polygon. If false, delete the tracks outside the polygon", required=True)
```
Comment (niravi) [inline scripts/polygon_cleanup_tracks.py]: What do you think of the name of this parameter? I don't really like it

File: scripts/polygon_cleanup_tracks.py (hunk 1 above)
Comment (apulverizer) [reply scripts/polygon_cleanup_tracks.py]: Yes, but it should be able to return in any supported CRS

File: scripts/polygon_cleanup_tracks.py (hunk 3 above)
Comment (apulverizer) [reply scripts/polygon_cleanup_tracks.py]: Hmm what about something like `-mode intersection/symetric-difference` ? Default would be intersection

**Comment by apulverizer** [comment]
@niravi did you test this against both Enterprise and Online. Just want to make sure the symmetric difference works for both.

**Comment by niravi** [comment]
@apulverizer fixed for Enterprise by creating an outside ring using the min and max values of proj 3857

File: polygon_cleanup_tracks.md [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+----
+
+Other than the authentication arguments (username, password, org) the script uses the following parameters:
+
+- -layer-url <layer_url> - The feature service URL for your polygon feature layer with features that will be spatially intersected with track points. This is required.
+- -where <where_clause> - The where clause used to filter out only certain polygons for the spatial comparison (for example, OBJECTID > 1). Defaults to 1=1 (all features are used for comparison)
+- --symmetric-difference <symmetric_difference> - A parameter when provided, deletes features that fall outside the polygons. If not provided, delete track points inside the polygons. 
+
+Example Usage 1
+```python
```
Comment (apulverizer) [inline polygon_cleanup_tracks.md]: should be bash since it's a command line example

File: scripts/polygon_cleanup_tracks.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+    # Add the handlers to the root logger
+    logger.addHandler(sh)
+    return logger
+
+
+def form_donut(rings):
+    for ring in rings:
+        ring.reverse()
+    # append the max extent as the clockwise outside ring
+    rings.append([[-20037508.3427892, -20037508.3427892], [-20037508.3427892, 20037508.3427892], [20037508.3427892, 20037508.3427892], [20037508.3427892, -20037508.3427892], [-20037508.3427892, -20037508.3427892]])
```
Comment (apulverizer) [inline scripts/polygon_cleanup_tracks.py:58]: Can you confirm this is the max extent? Seems different from https://epsg.io/3857

**Review by apulverizer — APPROVED** [review]
Looks good, just a few minor comments.

File: polygon_cleanup_tracks.md (hunk 4 above)
Comment (niravi) [reply polygon_cleanup_tracks.md]: 👍

File: scripts/polygon_cleanup_tracks.py (hunk 5 above)
Comment (niravi) [reply scripts/polygon_cleanup_tracks.py:58]: Technically it's `6378137 * Pi` which is what you see above. There's a comment on that link saying that the data is wrong

**Pull request merged**